<a href="https://colab.research.google.com/github/dhruvbhatnagar9548/DS_and_ML_projects/blob/main/Car_Sale_Data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# Exploring e-bay sales of car data
We will be working on a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to Kaggle. The version of the dataset we are working with is a sample of 50,000 data points that was prepared by Dataquest including simulating a less-cleaned version of the data.

The data dictionary provided with data is as follows:

- dateCrawled - When this ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which which year the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which which year the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings.

In [ ]:
import numpy as np
import pandas as pd
autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [ ]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [ ]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [ ]:
autos.describe()

yearOfRegistration       powerPS  monthOfRegistration  nrOfPictures  \
count        50000.000000  50000.000000         50000.000000       50000.0   
mean          2005.073280    116.355920             5.723360           0.0   
std            105.712813    209.216627             3.711984           0.0   
min           1000.000000      0.000000             0.000000           0.0   
25%           1999.000000     70.000000             3.000000           0.0   
50%           2003.000000    105.000000             6.000000           0.0   
75%           2008.000000    150.000000             9.000000           0.0   
max           9999.000000  17700.000000            12.000000           0.0   

         postalCode  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

In [ ]:
# Clean Columns
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [ ]:
#changing some columns names from real datset to make our understanding better towards the column
#Also, changing the columns from camelcase to snakecase

autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gear_box', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']

In [ ]:
# exploring the data
autos.describe(include = "all")

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-21 20:37:19  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year gear_box      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

In [ ]:
autos.head(2)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   

  offer_type   price  ab_test vehicle_type  registration_year   gear_box  \
0    Angebot  $5,000  control          bus               2004    manuell   
1    Angebot  $8,500  control    limousine               1997  automatik   

   power_ps   model   odometer  registration_month fuel_type    brand  \
0       158  andere  150,000km                   3       lpg  peugeot   
1       286     7er  150,000km                   6    benzin      bmw   

  unrepaired_damage           ad_created  num_photos  postal_code  \
0              nein  2016-03-26 00:00:00           0        79588   
1              nein  2016-04-04 00:00:00           0        71034   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08

In [ ]:
autos["gear_box"].value_counts()

manuell      36993
automatik    10327
Name: gear_box, dtype: int64

In [ ]:
autos["unrepaired_damage"].value_counts()

nein    35232
ja       4939
Name: unrepaired_damage, dtype: int64

In [ ]:
autos["vehicle_type"].value_counts()

limousine     12859
kleinwagen    10822
kombi          9127
bus            4093
cabrio         3061
coupe          2537
suv            1986
andere          420
Name: vehicle_type, dtype: int64

In [ ]:
autos["num_photos"].value_counts()

0    50000
Name: num_photos, dtype: int64

In [ ]:
autos["fuel_type"].value_counts()

benzin     30107
diesel     14567
lpg          691
cng           75
hybrid        37
andere        22
elektro       19
Name: fuel_type, dtype: int64

In [ ]:
autos["brand"].value_counts()

volkswagen        10687
opel               5461
bmw                5429
mercedes_benz      4734
audi               4283
ford               3479
renault            2404
peugeot            1456
fiat               1308
seat                941
skoda               786
mazda               757
nissan              754
smart               701
citroen             701
toyota              617
sonstige_autos      546
hyundai             488
volvo               457
mini                424
mitsubishi          406
honda               399
kia                 356
alfa_romeo          329
porsche             294
suzuki              293
chevrolet           283
chrysler            181
dacia               129
daihatsu            128
jeep                110
subaru              109
land_rover           99
saab                 80
daewoo               79
trabant              78
jaguar               77
rover                69
lancia               57
lada                 31
Name: brand, dtype: int64

In [ ]:
autos["seller"].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [ ]:
autos["offer_type"].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

Intial take on Dataset
- price and odometer are strings type need to convert into int
- The seller, offer_type, num_photos does not give much info can be dropped from orignal dataset
- There are few columns which can be converted from german to english

In [ ]:
autos = autos.drop(["seller", "offer_type", "num_photos"],axis = 1)
autos.shape

(50000, 17)

In [ ]:
autos["odometer"] = (autos["odometer"].str.replace("km","")
                     .str.replace(",","").astype(int))
autos.rename({"odometer":"odometer_km"}, axis=1, inplace=True)
autos["odometer_km"].head(2)

0    150000
1    150000
Name: odometer_km, dtype: int64

In [ ]:
autos["price"] = (autos["price"].str.replace("$","")
                  .str.replace(",","").astype(int))
autos.rename({"price":"price($)"}, axis=1, inplace=True)
autos["price($)"].head(10)              
                 

0    5000
1    8500
2    8990
3    4350
4    1350
5    7900
6     300
7    1990
8     250
9     590
Name: price($), dtype: int64

In [ ]:
#lets explore the column of odometer and price
#odometer column
autos["odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [ ]:
# pric column
print(autos["price($)"].describe())
print(autos["price($)"].value_counts().sort_index().head(20))
print(autos["price($)"].value_counts().sort_index(ascending=True).tail(20))
print(autos["price($)"].unique().shape)

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price($), dtype: float64
0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price($), dtype: int64
197000      1
198000      1
220000      1
250000      1
259000      1
265000      1
295000      1
299000      1
345000      1
350000      1
999990      1
999999      2
1234566     1
1300000     1
3890000     1
10000000    1
11111111    2
12345678    3
27322222    1
99999999    1
Name: price($), dtype: int64
(2357,)


In [ ]:
# There are 2357 unique values of prices 
# If we see that there are outliers present in the data for lower and upper prices 
# It can be seen that minimum values is 0 and maximum values is 1 Million
# so we will take only values between (1000 to 100000)

autos = autos[autos["price($)"].between(1000,100000)]
print(autos["price($)"].describe())

# so minimum values is now 1000 and maximum values is 99900

count    38587.000000
mean      7089.045482
std       7893.814349
min       1000.000000
25%       2200.000000
50%       4300.000000
75%       8950.000000
max      99900.000000
Name: price($), dtype: float64


In [ ]:
# Now explore the date columns 
autos[["date_crawled", "ad_created", "last_seen"]][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [ ]:
# It can be observe that the dates are stored as strings
(autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index())

2016-03-05    0.025527
2016-03-06    0.013891
2016-03-07    0.035167
2016-03-08    0.032602
2016-03-09    0.032472
2016-03-10    0.033353
2016-03-11    0.032835
2016-03-12    0.037396
2016-03-13    0.016016
2016-03-14    0.036670
2016-03-15    0.033612
2016-03-16    0.029077
2016-03-17    0.030477
2016-03-18    0.012802
2016-03-19    0.035167
2016-03-20    0.038173
2016-03-21    0.037266
2016-03-22    0.032420
2016-03-23    0.032213
2016-03-24    0.029025
2016-03-25    0.030528
2016-03-26    0.033094
2016-03-27    0.031384
2016-03-28    0.035349
2016-03-29    0.034001
2016-03-30    0.033016
2016-03-31    0.031410
2016-04-01    0.034571
2016-04-02    0.036333
2016-04-03    0.039132
2016-04-04    0.036904
2016-04-05    0.013346
2016-04-06    0.003265
2016-04-07    0.001503
Name: date_crawled, dtype: float64

In [ ]:
(autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index())

2015-06-11    0.000026
2015-08-10    0.000026
2015-09-09    0.000026
2015-11-10    0.000026
2015-12-30    0.000026
                ...   
2016-04-03    0.039443
2016-04-04    0.037318
2016-04-05    0.011973
2016-04-06    0.003369
2016-04-07    0.001322
Name: ad_created, Length: 74, dtype: float64

In [ ]:
(autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index())

2016-03-05    0.001088
2016-03-06    0.003576
2016-03-07    0.004561
2016-03-08    0.006246
2016-03-09    0.008889
2016-03-10    0.009822
2016-03-11    0.011740
2016-03-12    0.022210
2016-03-13    0.008397
2016-03-14    0.011999
2016-03-15    0.015005
2016-03-16    0.015446
2016-03-17    0.026408
2016-03-18    0.007360
2016-03-19    0.014616
2016-03-20    0.019799
2016-03-21    0.019670
2016-03-22    0.020810
2016-03-23    0.017908
2016-03-24    0.018504
2016-03-25    0.017778
2016-03-26    0.016016
2016-03-27    0.014072
2016-03-28    0.019411
2016-03-29    0.020784
2016-03-30    0.023479
2016-03-31    0.022754
2016-04-01    0.023220
2016-04-02    0.024905
2016-04-03    0.024438
2016-04-04    0.023402
2016-04-05    0.130977
2016-04-06    0.234742
2016-04-07    0.139969
Name: last_seen, dtype: float64

In [ ]:
import datetime as dt

columns = ["date_crawled","ad_created","last_seen"]

for date in columns:
    min_value = dt.datetime.strptime(autos[date].str[:10].min(), "%Y-%m-%d")
    min_value_1 = min_value.strftime("%Y-%m-%d")
    max_value = dt.datetime.strptime(autos[date].str[:10].max(), "%Y-%m-%d")
    max_value_2 = max_value.strftime("%Y-%m-%d")
    print(f"The minimum value of the column {date} is {min_value_1}")
    print(f"The maximum value of the column {date} is {max_value_2}")
    print("The diffrence in days are {}".format(str(max_value-min_value)[:7]))
    print("\n")

The minimum value of the column date_crawled is 2016-03-05
The maximum value of the column date_crawled is 2016-04-07
The diffrence in days are 33 days


The minimum value of the column ad_created is 2015-06-11
The maximum value of the column ad_created is 2016-04-07
The diffrence in days are 301 day


The minimum value of the column last_seen is 2016-03-05
The maximum value of the column last_seen is 2016-04-07
The diffrence in days are 33 days




It can be understood that
- date_crawled was done for 33 days
- The ad_created are from 301 days(~10 months)
- The last_seen alling with date_crawled

In [ ]:
autos["registration_year"].describe()

count    38587.000000
mean      2005.680125
std         86.727780
min       1000.000000
25%       2001.000000
50%       2005.000000
75%       2009.000000
max       9999.000000
Name: registration_year, dtype: float64

In [ ]:
# it can be seen that year are off can be look min(1000) and max(9999) values
# so I am considering only registered car between 1900 and 2016
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize = True).head(20)

# It appears that most of the veichle registered in the past 20 years

2005    0.074932
2006    0.071327
2004    0.070170
2003    0.066645
2007    0.060753
2008    0.059246
2002    0.057443
2009    0.055856
2001    0.055533
2000    0.053811
1999    0.046385
2011    0.043479
2010    0.042511
2012    0.035058
1998    0.034439
2013    0.021282
1997    0.021148
2014    0.017408
2016    0.017354
1996    0.014583
Name: registration_year, dtype: float64

In [ ]:
# price by brand
l = autos["brand"].value_counts(normalize=True)
l*100

volkswagen        21.102053
bmw               12.540695
mercedes_benz     11.163129
audi               9.766729
opel               8.916512
ford               5.873490
renault            3.731805
peugeot            2.792800
fiat               2.109398
skoda              1.907606
seat               1.730029
smart              1.662765
toyota             1.463664
mazda              1.425996
citroen            1.391019
nissan             1.364113
mini               1.089676
hyundai            1.076224
sonstige_autos     1.033175
volvo              0.898647
kia                0.769500
honda              0.734523
mitsubishi         0.688783
porsche            0.678021
chevrolet          0.661877
alfa_romeo         0.624210
suzuki             0.573089
dacia              0.328248
chrysler           0.317486
jeep               0.277128
land_rover         0.263675
jaguar             0.185649
subaru             0.172196
daihatsu           0.169505
saab               0.137219
daewoo             0

In [ ]:
# There a 4 out of 5 companies cars from German manufactures which covers approx. 50% of the market.
# In that Volkswagen is th company whose sale is roughly double than the sale of next two brands
#There are various brands which doen not have significant sale so we can drop those whose sale is below 5%
brands_counts = autos["brand"].value_counts(normalize=True)
common_brand = brands_counts[brands_counts>.05].index
print(common_brand)

Index(['volkswagen', 'bmw', 'mercedes_benz', 'audi', 'opel', 'ford'], dtype='object')


In [ ]:
mean_prices_of_brand = {}

for brand in common_brand:
    brands = autos[autos["brand"] == brand]
    mean_price = brands["price($)"].mean()
    mean_prices_of_brand[brand] = int(mean_price)
mean_prices_of_brand

{'volkswagen': 6645,
 'bmw': 8975,
 'mercedes_benz': 9196,
 'audi': 10276,
 'opel': 4219,
 'ford': 5274}

- It can be observe that BMW, AUDI, MERCEDES BENZ are costly comparing from other brands
- OPEL & FORD are cheap
- While brand whose sale are high as comparison to other brands price is in between

In [ ]:
# Now I will try to show the mileage comparison between the top brands
bmp_series = pd.Series(mean_prices_of_brand)
pd.DataFrame(bmp_series, columns=["mean_price"])

mean_price
volkswagen           6645
bmw                  8975
mercedes_benz        9196
audi                10276
opel                 4219
ford                 5274

In [ ]:
brand_mean_mileage = {}

for brand in common_brand:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)
print(brand_mean_mileage)
mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(mean_prices_of_brand).sort_values(ascending=False)

{'volkswagen': 125765, 'bmw': 132068, 'mercedes_benz': 130130, 'audi': 127524, 'opel': 123952, 'ford': 119651}


In [ ]:
brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info

mean_mileage
bmw                  132068
mercedes_benz        130130
audi                 127524
volkswagen           125765
opel                 123952
ford                 119651

In [ ]:
brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
bmw                  132068        8975
mercedes_benz        130130        9196
audi                 127524       10276
volkswagen           125765        6645
opel                 123952        4219
ford                 119651        5274

There is not much diffrence in the mileage of the brand as comparison to price. However we can observe that expensive veichle brand has a slightly better mileage less expensive veichles